### PR TITLE
plan: implement convert2NewPhysicalPlan for proj and sort

### PR DESCRIPTION
**Pull request by hanfei1991**
This is the first step of new plan builder.
I will implement datasource and selection in next pr and add tests for plan.

@shenli @coocood @zimulala PTAL

### CODE REVIEW

File: plan/plan.go [hunk 1]
```diff
@@ -148,17 +173,38 @@ type PhysicalPlan interface {
 
 	// Copy copies the current plan.
 	Copy() PhysicalPlan
+
+	// attach2TaskProfile makes the current physical plan as the father of task's physicalPlan and updates the cost of
+	// current task. If the child's task is cop task, some operator may close this task and return a new rootTask.
+	attach2TaskProfile(...taskProfile) taskProfile
```
Comment (coocood) [inline plan/plan.go]: When do we need to attach to multiple profile?

File: plan/task_profile.go [hunk 2]
```diff
@@ -77,21 +79,35 @@ func attachPlan2Task(p PhysicalPlan, t taskProfile) taskProfile {
...
-	t.cst += float64(t.cnt) * netWorkFactor
-	t.addPlan2Index = false
+	if t.addPlan2Index {
+		t.cst += float64(t.cnt) * netWorkFactor
+		t.addPlan2Index = false
+	}
 }
 
-func (t *copTaskProfile) finishTask() {
+func (t *copTaskProfile) finishTask(ctx context.Context, allocator *idAllocator) taskProfile {
```
Comment (coocood) [inline plan/task_profile.go]: Need comments of this function.

File: plan/task_profile.go [hunk 3]
```diff
@@ -63,8 +66,7 @@ func (t *copTaskProfile) copy() taskProfile {
 }
 
 func attachPlan2Task(p PhysicalPlan, t taskProfile) taskProfile {
```
Comment (coocood) [inline plan/task_profile.go]: s/attachPan2Task/attachPlan2TaskProfile

To keep it consistent with the method in Plan.

File: plan/plan.go (hunk 1 above)
Comment (hanfei1991) [inline plan/plan.go]: Join and Union

File: plan/physical_plans.go [hunk 4]
```diff
@@ -68,6 +68,55 @@ var (
...
+// Copy implements the PhysicalPlan Copy interface.
+func (p *PhysicalIndexReader) Copy() PhysicalPlan {
+	np := *p
+	np.basePlan = p.basePlan.copy()
+	np.basePhysicalPlan = newBasePhysicalPlan(np.basePlan)
+	return &np
+}
+
+// PhysicalIndexLookUpReader is the index look up reader in tidb.
+type PhysicalIndexLookUpReader struct {
```
Comment (shenli) [inline plan/physical_plans.go]: Where is this pla used?

File: plan/plan.go [hunk 5]
```diff
@@ -68,6 +68,25 @@ func (c *columnProp) equal(nc *columnProp, ctx context.Context) bool {
 	return c.col.Equal(nc.col, ctx) && c.desc == nc.desc
 }
 
+type requiredProp struct {
```
Comment (shenli) [inline plan/plan.go]: Add comment for this. Its name looks like `requiredProperty`.

File: plan/physical_plans.go (hunk 4 above)
Comment (hanfei1991) [inline plan/physical_plans.go]: double read

File: plan/task_profile.go [hunk 6]
```diff
@@ -156,33 +262,35 @@ func (sel *Selection) attach2TaskProfile(profiles ...taskProfile) taskProfile {
...
 		}
 		// TODO: Estimate t.cnt by histogram.
 		t.cnt = uint64(float64(t.cnt) * selectionFactor)
 	case *rootTaskProfile:
 		t.cst += float64(t.cnt) * cpuFactor
 		t.cnt = uint64(float64(t.cnt) * selectionFactor)
+		attachPlan2TaskProfile(sel.Copy(), t)
 	}
-	return profile
+	return nil
```
Comment (shenli) [inline plan/task_profile.go]: Why return nil here?

File: plan/task_profile.go [hunk 7]
```diff
@@ -156,33 +262,35 @@ func (sel *Selection) attach2TaskProfile(profiles ...taskProfile) taskProfile {
...
-			conditions = append(conditions[:i], conditions[i+1:]...)
+			tableConds = append(tableConds, cond)
+		} else {
+			indexConds = append(indexConds, cond)
 		}
 	}
-	if len(conditions) != 0 {
-		indexSel = sel
+	if len(indexConds) != 0 {
+		indexSel = Selection{Conditions: indexConds}.init(sel.allocator, sel.ctx)
```
Comment (shenli) [inline plan/task_profile.go]: Why create a new object?

File: plan/task_profile.go [hunk 8]
```diff
@@ -44,7 +47,7 @@ func (t *copTaskProfile) count() uint64 {
 	return t.cnt
 }
 
-func (t *copTaskProfile) setCost(cst float64) {
+func (t *copTaskProfile) addCost(cst float64) {
```
Comment (hanfei1991) [inline plan/task_profile.go]: well, This function should be t.cst += cst, my fault.

**Comment by hanfei1991** [comment]
@shenli @winoros PTAL

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**